ruby_programming/advanced_ruby/blocks.md

```diff
@@ -287,30 +295,40 @@ a_lambda.call
 # => 1
 ~~~
 
-A Proc object returns from the context in which it is called. If you are in the top level context (outside of a class and method) then you'll get an error because you can't return out of the very top level context, as there is no caller to return to.
+In non-lambda procs, `return` means exit from embracing method in which it was *defined* and will throw `LocalJumpError` if invoked outside the method. Check the [documentation](https://ruby-doc.org/core-2.7.0/Proc.html#class-Proc-label-Lambda+and+non-lambda+semantics) about it. In the top level context (not inside any method) a `return` will exit the program.
+
+Calling the proc (with an explicit `return`) will return from the method it was defined in:
 
 ~~~ruby
-a_proc = Proc.new { return }
+def execute_proc(proc_inside_method)
+  puts proc_inside_method.call # will produce no output!
+end
 
-a_proc.call
+def define_proc
+  proc_inside_method = proc { return 'explicit return' }
+  puts 'Calling the proc inside "execute_proc" will return from "define_proc".'
+  execute_proc(proc_inside_method)
+  puts 'This line is never reached.'
+end
 
-# => localJumpError (unexpected return)
+define_proc
 ~~~
 
-If you try the above in on repl.it you won't get an error, which is just to do with how they manage the context of code. If you try it in irb then you'd get the expected error.
+This behaviour might be a bit unintuitive (One might expect a return from `execute_proc`, where the proc was called.). So use an explicit return from a proc with care. This is an example in ruby, where it makes a difference, if you use an explicit or an implicit return. Try and change the definition of the proc to `proc { 'implicit return' }` in the above example. What happens?
 
-If you return from a proc inside a method, the method is the context in which it was called and therefore it returns from the method before any of the other code below it is executed.
+If calling the proc (with an explicit `return`) after the the method it was defined in, it will throw the expected `LocalJumpError`:
 
 ~~~ruby
-def my_method
-  a_proc = Proc.new { return }
-  a_proc.call
-  puts "this line is never reached"
+def define_proc
+  proc { return 'explicit return' }
 end
 
-my_method
+define_proc.call # calling the proc, after define_proc exited
```

Comment:
We would need to make clear here that the proc is implicitly returned as the last evaluated code in the method. It might seem obvious but when learning a new complex topic the fundamentals are often quickly forgotten. Just a comment saying the proc is implicitly returned would make it clearer.